db/compaction/compaction.cc

```diff
@@ -991,4 +998,83 @@ int Compaction::EvaluatePenultimateLevel(
   return penultimate_level;
 }
 
+void Compaction::FilterInputsForCompactionIterator() {
+  assert(earliest_snapshot_.has_value());
+  // cfd_ is not populated at Compaction construction time, get it from
+  // VersionStorageInfo instead.
+  assert(input_vstorage_);
+  const auto* ucmp = input_vstorage_->user_comparator();
+  assert(ucmp);
+  // Simply comparing file boundaries when user-defined timestamp is defined
+  // is not as safe because we need to also compare timestamp to know for
+  // sure. Although entries with higher timestamp is also supposed to have
+  // higher sequence number for the same user key (without timestamp).
+  assert(ucmp->timestamp_size() == 0);
+  std::vector<std::tuple<Slice, Slice, SequenceNumber>>
+      standalone_range_tombstones_for_filtering;
+  const LevelFilesBrief* flevel_start = input_levels(0);
+  for (size_t i = 0; i < flevel_start->num_files; i++) {
+    FileMetaData* fmeta = flevel_start->files[i].file_metadata;
+    if (!fmeta->FileIsStandAloneRangeTombstone()) {
+      continue;
+    }
+    bool range_del_in_snapshot =
+        ((fmeta->fd.smallest_seqno) <= (earliest_snapshot_.value()) &&
+         (snapshot_checker_ == nullptr ||
+          LIKELY(
+              snapshot_checker_->CheckInSnapshot(
+                  (fmeta->fd.smallest_seqno), (earliest_snapshot_.value())) ==
+              SnapshotCheckerResult::kInSnapshot)));
+    if (!range_del_in_snapshot) {
+      continue;
+    }
+    standalone_range_tombstones_for_filtering.emplace_back(
+        fmeta->smallest.user_key(), fmeta->largest.user_key(),
+        fmeta->fd.smallest_seqno);
+  }
+  if (standalone_range_tombstones_for_filtering.empty()) {
+    return;
+  }
+
+  std::vector<std::vector<FileMetaData*>> level_files;
+  size_t num_input_levels = input_levels_.size();
+  level_files.reserve(num_input_levels - 1);
+  for (size_t level = 1; level < num_input_levels; level++) {
```

Comment:
Does this handle L0 range tombstone file covering another L0?

Reply:
That's a great point! @pdillinger thank you spotting this and saved me from my stress test debug hell.  I have added a TODO for the possible future improvement here.